### ISSUE TITLE
tracing the model for deploying Keybert on Aws Inferentia

### ISSUE THREAD

**Issue by DevBey** [MaartenGr/KeyBERT#81]
Hi, 

we have been using Keybert in for quite some time now, while currently its' deployed in AWS sagemaker GPU instance which is very costly .

We want to move this to aws inferentia chips which are way faster than t4 GPU and considerably cheap. all our pytorch models are deployed in aws inferentia now except this.

for tracing we use torch.jit.trace, please let me know how could we do same with keybert !!

**Comment by MaartenGr** [comment]
I am not familiar with the AWS inferentia chips, are you running into any specific issues using it together with KeyBERT? From my understanding, you can use the same tracing between both back-ends.

**Comment by DevBey** [comment]
Hi @MaartenGr,

Thanks for such a prompt response.

Yes I was able to deploy this on AWS inferentia with 90 percent reduced cost and faster inference than t4.

I had already traced "mpnet" sentence Transformer model, and just loaded it into Keybert class in a similar way.

Before I close this, I think it would be great we can use keybert using huggingace models,Autotokenizer etc with a few functions for preprocessing and post using Keybert functions, this will make the keybert project reach out to a lot of people.

Anyway, Kudos to the whole Keybert Team.

**Comment by MaartenGr** [comment]
Glad to hear that it works well! 

With respect to the huggingface models, you can already implement them using the `Flair` library which is typically a bit more intuitive to use for those starting out with language models. Huggingface will be directly supported in the future. However, you can use a [custom backend](https://maartengr.github.io/KeyBERT/guides/embeddings.html#custom-backend) if you need any additional options. 

With respect to the preprocessing and postprocessing, are you referring to the documents themselves or the resulting keywords? What do you have specifically in mind?

**Comment by DevBey** [comment]
Hi @MaartenGr,

https://maartengr.github.io/KeyBERT/guides/embeddings.html#custom-backend this is brilliant, I have been using flair and hugging face for quite a few years now.

So what I had in my mind was supporting huggingFace directly and letting people use the keybert from all sorts of huggingface models instead of having to install other libraries !!!

for example, we could directly use sentence transformer with huggingface -> 

from transformers import AutoTokenizer, AutoModel
import torch
import torch.nn.functional as F

#Mean Pooling - Take attention mask into account for correct averaging
def mean_pooling(model_output, attention_mask):
    token_embeddings = model_output[0] #First element of model_output contains all token embeddings
    input_mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size()).float()
    return torch.sum(token_embeddings * input_mask_expanded, 1) / torch.clamp(input_mask_expanded.sum(1), min=1e-9)


# Sentences we want sentence embeddings for
sentences = ['This is an example sentence', 'Each sentence is converted']

# Load model from HuggingFace Hub
tokenizer = AutoTokenizer.from_pretrained('sentence-transformers/all-mpnet-base-v2')
model = AutoModel.from_pretrained('sentence-transformers/all-mpnet-base-v2')

# Tokenize sentences
encoded_input = tokenizer(sentences, padding=True, truncation=True, return_tensors='pt')

# Compute token embeddings
with torch.no_grad():
    model_output = model(**encoded_input)

# Perform pooling
sentence_embeddings = mean_pooling(model_output, encoded_input['attention_mask'])

# Normalize embeddings
sentence_embeddings = F.normalize(sentence_embeddings, p=2, dim=1)

print("Sentence embeddings:")
print(sentence_embeddings)


----------------------------------------------------------------------------------

if similarly, we could have a page in hugging face regarding keybert and some basic tutorial on how to use it with and without Keybert library. I think that would really help in popularizing and letting more people access it !!

**Comment by MaartenGr** [comment]
Agreed! It would be nice to implement huggingface as an additional dependency from which you can directly call one of the models on https://huggingface.co/models. Thus far, I have been quite hesitant to implement it as there can be quite a few differences between models which make it difficult to develop a single way of creating the embeddings for each model. 

To give you an example, the procedure that you sketch above is specific for sentence-transformers whereas other models are not specifically trained and/or optimized for the steps you mentioned above, like pooling and normalization. In practice, I do not believe it should be an issue but some extensive tests across models should be preferred. Do you think the steps above make sense for all/most huggingface models?

**Comment by MaartenGr** [comment]
Huggingface transformers was implemented a while ago, so I will go ahead and close this for now. However, if you want to continue this discussion, please let me know and I'll make sure to re-open this.
